### PR TITLE
Feat/add tests

### PR DESCRIPTION
**Pull request by qplevier**

### CODE REVIEW

File: metadata_automation/cli.py [hunk 1]
```diff
@@ -416,9 +416,30 @@ def shacl_from_shaclplay(
...
+    type=click.Path(),
+    default="./outputs/sempyro_classes",
+    help="Output directory for SeMPyRO Pydantic classes.",
+)
+@click.option(
+    "--imports-path",
+    type=click.Path(exists=True),
+    default="./inputs/sempyro/imports.yaml",
+    help="Path to imports configuration YAML file.",
+)
```
Comment (agjharms) [inline metadata_automation/cli.py:436]: Can you update the Readme to document the CLI changes?

File: tests/test_cli_shaclplay.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+        for sheet_name in expected_sheets:
+            actual_df = pd.read_excel(actual_path, sheet_name=sheet_name, header=None)
+            expected_df = pd.read_excel(expected_path, sheet_name=sheet_name, header=None)
+
+            if sheet_name == "NodeShapes (classes)":
+                if actual_df.shape[0] > 6 and actual_df.shape[1] > 1:
+                    actual_df.iat[6, 1] = None
+                if expected_df.shape[0] > 6 and expected_df.shape[1] > 1:
+                    expected_df.iat[6, 1] = None
```
Comment (agjharms) [inline tests/test_cli_shaclplay.py:32]: What do these lines do?

File: tests/test_cli_shacl_from_shaclplay.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        assert "No SHACLPlay Excel files found" in result.output
+
+    def test_shacl_from_shaclplay_invalid_excel(self, runner, test_input_dir, tmp_path):
+        """Test error handling for invalid Excel structure."""
+        shaclplay_dir = tmp_path / "shaclplay"
+        shaclplay_dir.mkdir()
+
+        excel_path = test_input_dir / "invalid_shaclplay.xlsx"
+        target_path = shaclplay_dir / "SHACL-invalid.xlsx"
+        target_path.write_bytes(excel_path.read_bytes())
```
Comment (agjharms) [inline tests/test_cli_shacl_from_shaclplay.py:84]: Why did you `write_bytes` instead of copying it using `shutil`, for example?

File: tests/test_cli_error_handling.py [hunk 4]
```diff
@@ -0,0 +1,596 @@
...
+            mock_resolved = MagicMock()
+
+            mock_file_path.parent = mock_parent1
+            mock_parent1.parent = mock_parent2
+            mock_parent2.resolve.return_value = mock_resolved
+            mock_resolved.__truediv__ = lambda self, other: mock_template
+
+            real_path = Path
+
+            def path_side_effect(arg):
```
Comment (agjharms) [inline tests/test_cli_error_handling.py]: Unless there is a pressing need for a different structure, move all function definitions to the top of the test.

File: tests/README.md [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+reporting), and ruff formatting warnings.
+
+### Unit tests for utility modules (`test_coverage_units.py`)
+
+Unit tests target specific utility functions and edge cases in modules such as the LinkML creator,
+SeMPyRO helpers, SHACLPlay converter and utilities, and vocabulary mappings. They focus on YAML
+loading and mutation, import parsing, validation logic application, RDF model wiring, SHACLPlay
+Excel creation, property conversion branches, and custom Pydantic generation.
+
+**Total:** 69 tests (see pytest output)
```
Comment (agjharms) [inline tests/README.md]: Don't include the total number of tests; it will be outdated quite fast.

File: tests/test_coverage_units.py [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+def test_shaclplay_utils() -> None:
+    assert slugify_property_label("Access Rights!") == "access-rights"
+    assert parse_cardinality("1") == (1, 1)
+    assert parse_cardinality("0..n") == (None, None)
+    assert parse_cardinality("1..n") == (1, None)
+    assert parse_cardinality("0..1") == (None, 1)
+    assert parse_cardinality("") == (None, None)
+
+    now_iso = get_current_datetime_iso()
+    assert len(now_iso) >= 19
```
Comment (agjharms) [inline tests/test_coverage_units.py]: Split this up into a test per function.

File: tests/test_coverage_units.py [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+
+def test_shaclplay_utils() -> None:
+    assert slugify_property_label("Access Rights!") == "access-rights"
+    assert parse_cardinality("1") == (1, 1)
+    assert parse_cardinality("0..n") == (None, None)
+    assert parse_cardinality("1..n") == (1, None)
+    assert parse_cardinality("0..1") == (None, 1)
+    assert parse_cardinality("") == (None, None)
+
+    now_iso = get_current_datetime_iso()
```
Comment (agjharms) [inline tests/test_coverage_units.py]: A more appropriate way to test this is with a library like 'freezegun' as is done for img2catalog. But to be honest, `get_current_datetime_iso()` only calls 1 method from the datetime library; I would skip this test.

File: tests/test_cli_additional_coverage.py [hunk 8]
```diff
@@ -0,0 +1,548 @@
...
+import pytest
+from click.testing import CliRunner
+
+from metadata_automation.cli import main
+
+
+@pytest.fixture
+def runner():
+    """Create a Click CLI runner."""
+    return CliRunner()
```
Comment (agjharms) [inline tests/test_cli_additional_coverage.py]: The runner fixture is re-used a lot and can therefore be put in conftest.py.

File: tests/test_cli_additional_coverage.py [hunk 9]
```diff
@@ -0,0 +1,548 @@
...
+                    str(test_file),
+                    "-o",
+                    str(tmp_path / "output"),
+                ],
+            )
+
+        assert result.exit_code == 1
+        assert "Error: Failed to read classes sheet" in result.output
+
+    def test_missing_sheet_name_in_row(self, runner, tmp_path):
```
Comment (agjharms) [inline tests/test_cli_additional_coverage.py]: This and the `test_missing_class_uri_in_row` can probably be unified using a parametric test.

File: tests/test_cli_additional_coverage.py [hunk 10]
```diff
@@ -0,0 +1,548 @@
...
+                str(tmp_path),
+                "-o",
+                str(tmp_path / "output"),
+            ],
+        )
+
+        assert result.exit_code == 1
+        assert "Error: Could not extract namespace from SHACL-test.xlsx" in result.output
+
+    def test_subprocess_with_stdout_output(self, runner, tmp_path):
```
Comment (agjharms) [inline tests/test_cli_additional_coverage.py]: Is there a reason to test the stdout and stderr output seperately?

File: tests/test_utils.py
Comment (agjharms) [inline tests/test_utils.py:1]: Shouldn't this be 'test_utils.py'?

File: tests/test_cli_shacl_from_shaclplay.py [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+        nonexistent_dir = tmp_path / "nonexistent"
+
+        result = runner.invoke(
+            shacl_from_shaclplay, ["--input-path", str(nonexistent_dir), "--output-path", str(tmp_path / "output")]
+        )
+
+        assert result.exit_code != 0
+        assert "Invalid value for '-i' / '--input-path'" in result.output
+
+    def test_shacl_from_shaclplay_preserves_namespace(self, runner, shaclplay_input_dir, tmp_path):
```
Comment (agjharms) [inline tests/test_cli_shacl_from_shaclplay.py]: Doesn't this just test if the prefixes are copied over correctly by the SHACLPlay Java application?

File: tests/test_cli_sempyro.py [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+                str(test_excel),
+                "--namespace",
+                "hri",
+            ]
+            + cli_args_with_temp_paths,
+        )
+
+        # Should report how many classes were found in the Excel file
+        assert "✓ Found" in result.output
+        assert "classes in Excel file" in result.output
```
Comment (agjharms) [inline tests/test_cli_sempyro.py]: Does this test the accuracy of the counting? It only seems to test if any counting is happening.

File: tests/test_cli_shaclplay.py (hunk 2 above)
Comment (qplevier) [reply tests/test_cli_shaclplay.py:32]: This is the 7th row, 2nd column which is a dcterms:modified dynamic timestamp at which the actual_df is created (and the expected -> can be manually made empty in tests/tests_expected/[default|multi|namespace_override]/SHACL-[testclass|classa|classb].xlsx to remove 1 if statement per file). It needs to be set to none to allow for comparing the two files excluding the timestamps.

File: tests/README.md [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+```
+tests/
+├── conftest.py                         # Shared pytest fixtures
+├── test_cli_shaclplay.py              # Integration tests for shaclplay CLI command
+├── test_cli_shacl_from_shaclplay.py   # Integration tests for shacl_from_shaclplay CLI command
+├── test_cli_sempyro.py                # Integration tests for sempyro CLI command
+├── test_cli_error_handling.py         # Shared CLI error-handling tests across commands
+├── test_cli_additional_coverage.py    # Additional edge-case coverage for CLI commands
+├── test_coverage_units.py             # Unit tests for utility and helper modules
+├── create_test_files.py               # Script to regenerate test input files
```
Comment (agjharms) [inline tests/README.md]: This file seems to be missing.

File: tests/test_cli_shacl_from_shaclplay.py (hunk 3 above)
Comment (qplevier) [reply tests/test_cli_shacl_from_shaclplay.py:84]: no need for shutil import and the size of the test files is a few KB so can be done through memory instead of chuncks.

**Review by agjharms — APPROVED** [review]
